### ISSUE TITLE
Sticker packs

### ISSUE THREAD

**Issue by epoberezkin** [simplex-chat/simplex-chat#2320]

**Comment by PalmTamino** [comment]
If implementing stickers, directly think about

- Making them usable as emojis too in messages (Discord has for example)
- Animated stickers (Signal has for example)

Maybe use a format of stickers that already exists, so that not completely new packs have to be created, for example, use Signal Stickers.

**Comment by highschooldropout** [comment]
I'd rather have emoji reactions.

**Comment by PalmTamino** [comment]
> I'd rather have emoji reactions.

Well, thats a nice to have, but it wasn't a question if we want this or that?

**Comment by highschooldropout** [comment]
> > I'd rather have emoji reactions.
> 
> Well, thats a nice to have, but it wasn't a question if we want this or that?

Just figured I'd throw in my $0.02 as the two serve similar-enough functions. If not, maybe I'm confused about stickers...

**Comment by PalmTamino** [comment]
> Hi there, I love the concept of the app and love the suggestion! A must have feature for me that is keeping me on telegram is their stickers and gifs management. Having the option to make my own stickers and sync them alongside gifs I use is just a must have in my opinion...
> Would that be possible within simplex?

Exactly, same for me too, I just hope those stickers will work in text messages too, like as custom emojis.

**Comment by TGRush** [comment]
> > > I'd rather have emoji reactions.
> > 
> > Well, thats a nice to have, but it wasn't a question if we want this or that?
> 
> Just figured I'd throw in my $0.02 as the two serve similar-enough functions. If not, maybe I'm confused about stickers...

Stickers are re-usable images (or Videos) that can be used to express all kinds of things in chats. They usually have a limited resolution (e.g 512 pixels max) and are always shown in a set size.

Many people like to use them to bring a certain personal touch into conversations or to respond to certain things in a more dynamic way

Usually, they're ordered info "Sticker Packs", often to represent a certain interest (like sports), purpose (like reaction images) or character.

**Comment by epoberezkin** [comment]
what's wrong with GIPHY keyboard?

**Comment by PalmTamino** [comment]
> what's wrong with GIPHY keyboard?

Gifs are not a sticker replacement. Also I would like to use stickers in normal messages, like for discord, in the text. 

if you copy them, they can have a name like [Doge], [Facepalm],… the creator of the pack can set the name for what happens if u copy it and use outside the app.

**Comment by epoberezkin** [comment]
GIPHY has stickers too, both animated and not.

**Comment by guillaumeboehm** [comment]
It's actually very cumbersome to type whatever you want to express to find a good enough gif/sticker and maybe not find something for what you want, when you know you have the perfect sticker but can't use it, it's the case for me when I'm using things like whatsapp or such instead of Telegram.

**Comment by TGRush** [comment]
> what's wrong with GIPHY keyboard?

GIFs are a format with worse quality and GIPHY is a REALLY bad provider due to their privacy practices.

Stickers as implemented in Telegram, Signal, and many more software is able to leverage the strengths of PNG and WEBP and can be made completely private.

I encourage everyone here to read GIPHY's privacy policy. https://support.giphy.com/hc/en-us/articles/360032872931

**Comment by whereismybugfix** [comment]
I strongly support this feature request. For many users this may seem like a dealbreaker. It's understandable, stickers might be not the highest priority thing. Nonetheless, SimpleX chats feel quite empty without them.

**Comment by PalmTamino** [comment]
> Another point, which was implied before but not explicitly said, is that the structure of stickers --being organized into user-generated packs-- creates a specific type of organic person-to-person dissemination of sticker packs through their usage, which is not the case with something like GIPHY. I'm not a big sticker user, but I think this point is key to why they are so popular.

Exactly, I would even love if they can be used in messages, like in Discord or to some extent in WeChat, which has custom emojis. This makes conversations way more fun and brings more life to them as I noticed. 
![IMG_1227](https://github.com/simplex-chat/simplex-chat/assets/36358673/8744a345-2b4d-424f-92ef-bb2713b34c44)

**Comment by PalmTamino** [comment]
> OpenMoji 14.0.0
> 
> Open source emojis for designers, developers and everyone else!
> 
> https://openmoji.org/
> 
> Explore collection:
> 
> https://openmoji.org/library/

They look awful, we should either stick to system emojis or use the iOS ones like Signal does. I think there‘s no real copyright on fonts, thats why they‘re allowed to.

**Comment by epoberezkin** [comment]
yes, I am not a fan of openmoji too, and android emoji's are terrible frankly.

@PalmTamino you said you have a ux design studio and want to help with all that? Or maybe it was somebody else, sorry then :)

**Comment by Arturro43** [comment]
I would love to see custom sticker packs. I love Telegram's implementation:

- Using StickerBot, you create and manage sticker packs.

- Each sticker is assigned to an emoji (I don't like that)

- When sending a message that contains only some emoji Telegram suggests using a sticker from a saved stickerpack (I have a better idea how to do that)

- You can put a small "sticker" inside of your message using emojipacks (more on that later).

In my opinion, the best solution instead of assigning an emoji to a sticker would be to use some sort of identifier, for example $sticker_name$, this gives the author better way of defining what the sticker represents and potentially give it a unique name that can be easily remembered by sticker pack users who prefer to use "commands" instead of using the UI designed to select stickers. Unfortunately, this carries a theoretical possibility of de-anonymization, which I will mention in "Notes."

Regarding emojipacks:

I think they are just as important as stickers, because they allow for a much better expression of the feelings standing behind the message than classic emoji.

In this case, Discord's solution is the best - if you want to insert some emoji from a custom pack, just type :sticker_name:. Again, this carries the theoretical possibility of deanonymization, which I will mention in "Notes".

My idea for adding something unique that you can't find in any other messenger that would have a chance to make the conversation even more lively is to create "audio packs" - short audio files containing some sound effects, quotes, etc. In my country, there are plenty of extremely funny/legendary quotes that can be used for reaction purposes. In this case, I have no idea of the "syntax" "ID" of the sound effect. You would also have to come up with a catchier name than "sound effect" (maybe just "audio"?). In the future, you could add an "audio reaction" to a specific message using just the sounds from the audio packs, and also stickers with a sound file associated with them.

In conclusion, the best native solution for creating sticker/emoji/audio packs would be to create a section in the app settings called "Sticker & Emoji packs" or, for example, "Resource packs" or "Sticker, Emoji & Audio packs" (as long as the audio packs idea is a good;; I'll use the name "resource packs" from now on to avoid the constant repetition of "Sticker, Emoji & Audio packs"), where you choose if you want to create a sticker, emoji or audio pack, then you set the name of the resource pack, select an image/audio file from the files on your device and assign an "ID" to it, $sticker_name$ for stickers, :emoji_name: for custom emoji.

Sharing resource packs is also very important. The way it looks in Telegram feels very native. To add a particular pack to the "vault" of your sticker/emoji packs, just tap on the sticker/emoji someone sent you and it opens the add screen, where we can also view the contents of the pack you want.

Notes:

It is important that both stickers and emoji have a certain maximum size to prevent sending stickers that will be cut off/displayed all over the screen. This is where Telegram's implementation falls short, as quite often there are some memes used as stickers that become unreadable when scaled to a 512x512 square (one of the sides must be 512px and the other 512px or less). Therefore, the best solution would be to determine two sizes of stickers - the one displayed and the one displayed after pressing, which causes zoom.

Theoretically, using some sort of resource pack could contribute to the de-anonymization of their users. It's not a problem to have some sort of closed user community in which some member creates their own resource pack, and their members start using that resource pack. A potential observer could try to determine that a particular user is from a certain community (or at least potentially knows/uses a certain language) because they are using the same resource pack that the community in question was using (or the IDs in the resource pack are written in a particular language).
The solution to this problem would be:
- creation of a publicly available list of existing resource packs to achieve deniability (which I do not recommend, because there would certainly be resource packs containing controversial/banned content, see: banned photo of Putin in Russia, images used by, for example, protesters in Hong Kong, etc.; someone would have to deal with moderation of such a list and remove anything that could harm SimpleX - this solution only minimizes the risk of deanonymization, and involves additional work).
- not allowing users to create custom resource packs (this is a bad idea from the point of view of users who would like to express themselves and their emotions in their own original way) 
- notifying users that the use of resource packs other than official ones can, in theory, lead to their deanonymization (a better solution, but users may ignore/forget about it after some time, I think such a warning should also exist)
- introduction of a mechanism that creates unique copies of resource packs for each and every user - when a resource pack is saved, a 1:1 copy of it is created (preferably in a slightly modified form so that, for example, the hash of a photo/audio files is different from the original; I think this is the best option). Unfortunately, this brings up new issues:
  - Potential violation of copyrights (this can be resolved by informing the user who is creating the resource pack that by creating it, they are waiving their copyrights, as each person using their resource pack will have a unique copy of everything that is included in their resource pack)
  - since each user owns a unique copy of the resource pack (preferably with the name changed; which may be the n-th copy of the copy), and someone (who loves resource packs) owns a huge amount of them then after some time it is hard to remember exactly what resources you already own. Lets say you liked a sticker/emoji/audio that you didn't even know you already owned - how are you supposed to know that you already own a particular resource pack? (This can be solved by: 1. preventing users from adding more resources to the resource pack 2. preventing users from editing the "IDs" assigned to the resource pack (this way each copy will actually be based on an existing resource pack) in order to 3. be able to compare the existing IDs of the resource in the resource packs you already own. If specific IDs match (and, for example, there is a 100% similarity in the IDs occurring in one owned and a resource pack being added) then you can show the user prompt that he probably already owns a resource pack because it contains the same ID and indicate which one it is about and let user decide if he's sure he wants to add it to his "vault")
- additionally, you can introduce a requirement that each "ID" of the resource can contain only a-z characters, to minimize the risk of (still theoretical) deanonymization by assuming that a given user knows a certain language, because the ID of the resource contains special characters used only in a specific language ( unfortunately, this would lead to the introduction of a limitation of the ability to quickly choose a resource by using its ID for users who do not use languages based on Latin characters)

Introducing "autocomplete" for resources may raise the possibility that the user may think that the app (or less technical - that the creator of the resource pack) is keylogging. (You can let the user know how resource packs work and that everything happens on the device)

I can't wait to see stickers in SimpleX, it will certainly allow me to convince hundreds of friends to switch to SimpleX:)

**Comment by epoberezkin** [comment]
In a decentralised environment, the risk to deanonimization is not only via the names, but also via a different version of resource packs with some metadata embedded.

imagine if somebody wants to identify the users, they can simply send slightly different versions of the same pack to different users (would require client modification if the client sends the pack or of the server hosting the pack) - assuming people would want to be able to use the pack when they receive the sticker from that pack…

It is massively complex to do, and even figure out how to do without compromising privacy…

**Comment by Arturro43** [comment]
> In a decentralised environment, the risk to deanonimization is not only via the names, but also via a different version of resource packs with some metadata embedded.

Absolutely. My idea is to limit absolutely everything that can even remotely lead to deanonimization. I'll create new issue to discuss the best approach to this problem.
 
> imagine if somebody wants to identify the users, they can simply send slightly different versions of the same pack to different users

That's true. I think when it comes to group chats risk of tracking users by sending them "poisoned" resource packs (~unique copies of resouce packs) is much less possible, but it is still possible. Threat actor could observe exchange of "poisoned" resource packs between groups or in some (rare) cases even between individuals.

> It is massively complex to do, and even figure out how to do without compromising privacy…

Oh it is. Every single time when I think about what implementation would be the best to preserve user's privacy I get stopped by another problem. I think the easiest (and possibly the best?) way is to warn users about possible dangers of using custom resource packs.

**My UI proposition with added warning:**

<details>
![warning](https://github.com/simplex-chat/simplex-chat/assets/10798277/ea8a7ea5-eba3-4d63-a001-02e582c452b7)
</details>

**Comment by cybirdsec** [comment]
Just want to bump this thread after replying to [Feature]: Custom Sticker, Emoji & audio packs - your input needed #4014; this is most anticipated feature by far!
